Sharing GenBcode backend between scalac and dotty

DarkDimius opened the pull request:
This PR abstracts GenBcode implementation classes from existence of global using forwarders, which would have a different implementation in dotty.

The main class here is BackendInterface that defines everything that backend uses.
While it would be good to have backend use less that 150 methods on `Symbol` this commit doesn't address this issue: it tries to not introduce any changes in behavior.

In order to not loose perfromance and not allocate options during pattern matching this implemtation relies on name-based pattern matching with preallocated objects, this makes BackendInterface non-thread safe, but as of now it is used only in single thread. If several threads will ever be used to transfrom AST to asm bytecode than each of threads will need to own its own instance of BackendInterface.

@lrytz please have a look.

DarkDimius committed:
Factoring out methods used in backend. No direct usages of global anymore.

DarkDimius committed:
Reduced number of methods used in backend.

DarkDimius committed:
Factoring out definitions used in backend.

DarkDimius committed:
Starting to implement interfaces for scalac

DarkDimius committed:
Organise members of SymbolHelpers.

DarkDimius committed:
Use name-based pattern matching in backend

DarkDimius committed:
Make sure genIcode and genASM compile.

They are using primitives and opcodes that were extracted to objects.

DarkDimius committed:
Use name-based pattern matching for deconstructors, using same instances.

Saves allocations but breaks thread-safety of  BackendInterface.

DarkDimius committed:
Implement BackendInterface for scalac

DarkDimius committed:
Start using ScalacBackendInterface.

DarkDimius committed:
Fix initialization order problem.

DarkDimius committed:
ScalacBackend should now be functional.

DarkDimius commented:
I cannot reproduce error with which jenkins failed: https://scala-webapps.epfl.ch/jenkins/job/pr-scala-publish-core/4763/console
`scala.reflect.internal.FatalError: object ScalaRunTime does not have a member hash`
It does and on my local setup it can find it.
I'm not aware of internal specifics of how symbol loading in scalac is organized. @lrytz could you please have a look?

gkossakowski commented:
Did you try `ant all.clean && ant`?

DarkDimius commented:
I've tried making a clean clone and building it.
Now I've checked with `ant all.clean` still works(ie can't reproduce error).

retronym commented:
I reproduced it with:

```
% cat build.properties
#locker.skip=1
starr.use.released=1

% ant all.clean build-opt

init:
     [copy] Copying 2 files to /Users/jason/code/scala2/build/deps/junit
     [copy] Copying 6 files to /Users/jason/code/scala2/build/deps/partest
     [copy] Copying 1 file to /Users/jason/code/scala2/build/deps/repl
     [copy] Copying 5 files to /Users/jason/code/scala2/build/deps/scaladoc
     [echo] Using Scala 2.11.2 for STARR.
     [copy] Copying 5 files to /Users/jason/code/scala2/build/deps/starr
     [echo]        build time: 19 November 2014, 08:39:01
     [echo]      java version: Java HotSpot(TM) 64-Bit Server VM 1.8.0_25 (1.8)
     [echo]         java args: -Xmx3g -Xms3g -XX:+TieredCompilation -XX:ReservedCodeCacheSize=256m -XX:MaxPermSize=384m -XX:+UseNUMA -XX:+UseParallelGC
     [echo]        javac args:
     [echo]       scalac args: -feature  -optimise
     [echo] scalac quick args: -feature  -optimise
     [echo]          git date: 20141118-214623
     [echo]          git hash: 4c66397573
     [echo]     maven version: 2.11.5-SNAPSHOT
     [echo]      OSGi version: 2.11.5.v20141118-214623-4c66397573
     [echo] canonical version: 2.11.5-20141118-214623-4c66397573

asm.done:
    [mkdir] Created dir: /Users/jason/code/scala2/build/asm/classes
    [javac] Compiling 82 source files to /Users/jason/code/scala2/build/asm/classes
    [javac] warning: [options] bootstrap class path not set in conjunction with -source 1.5
    [javac] warning: [options] source value 1.5 is obsolete and will be removed in a future release
    [javac] warning: [options] To suppress warnings about obsolete options, use -Xlint:-options.
    [javac] Note: Some input files use unchecked or unsafe operations.
    [javac] Note: Recompile with -Xlint:unchecked for details.
    [javac] 3 warnings
[stopwatch] [asm.timer: 1.549 sec]
    [mkdir] Created dir: /Users/jason/code/scala2/build/libs

forkjoin.done:
    [mkdir] Created dir: /Users/jason/code/scala2/build/libs/classes/forkjoin
    [javac] Compiling 10 source files to /Users/jason/code/scala2/build/libs/classes/forkjoin
    [javac] warning: [options] bootstrap class path not set in conjunction with -source 1.5
    [javac] warning: [options] source value 1.5 is obsolete and will be removed in a future release
    [javac] warning: [options] To suppress warnings about obsolete options, use -Xlint:-options.
    [javac] 3 warnings
    [javac] Creating empty /Users/jason/code/scala2/build/libs/classes/forkjoin/scala/concurrent/forkjoin/package-info.class
      [jar] Building jar: /Users/jason/code/scala2/build/libs/forkjoin.jar
[stopwatch] [forkjoin.timer: 0.273 sec]

locker.start:

locker.lib:
    [mkdir] Created dir: /Users/jason/code/scala2/build/locker/classes/library
    [javac] Compiling 32 source files to /Users/jason/code/scala2/build/locker/classes/library
    [javac] warning: [options] bootstrap class path not set in conjunction with -source 1.5
    [javac] warning: [options] source value 1.5 is obsolete and will be removed in a future release
    [javac] warning: [options] To suppress warnings about obsolete options, use -Xlint:-options.
    [javac] 3 warnings
[locker.library] Compiling 579 files to /Users/jason/code/scala2/build/locker/classes/library
[locker.library] Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=384m; support was removed in 8.0
[locker.library] /Users/jason/code/scala2/src/library/scala/SerialVersionUID.scala:15: warning: Implementation restriction: subclassing Classfile does not
[locker.library] make your annotation visible at runtime.  If that is what
[locker.library] you want, you must write the annotation class in Java.
[locker.library] class SerialVersionUID(value: Long) extends scala.annotation.ClassfileAnnotation
[locker.library]       ^
[locker.library] warning: there were 120 deprecation warnings; re-run with -deprecation for details
[locker.library] warning: there was one unchecked warning; re-run with -unchecked for details
[locker.library] warning: there were 348 inliner warnings; re-run with -Yinline-warnings for details
[locker.library] four warnings found
[propertyfile] Creating new property file: /Users/jason/code/scala2/build/locker/classes/library/library.properties
[stopwatch] [locker.library.timer: 46.579 sec]

locker.reflect:
    [mkdir] Created dir: /Users/jason/code/scala2/build/locker/classes/reflect
[locker.reflect] Compiling 157 files to /Users/jason/code/scala2/build/locker/classes/reflect
[locker.reflect] Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=384m; support was removed in 8.0
[locker.reflect] warning: there were 60 deprecation warnings; re-run with -deprecation for details
[locker.reflect] warning: there were two unchecked warnings; re-run with -unchecked for details
[locker.reflect] warning: there were 74 inliner warnings; re-run with -Yinline-warnings for details
[locker.reflect] three warnings found
[propertyfile] Creating new property file: /Users/jason/code/scala2/build/locker/classes/reflect/reflect.properties
[stopwatch] [locker.reflect.timer: 46.396 sec]

locker.comp:
    [mkdir] Created dir: /Users/jason/code/scala2/build/locker/classes/compiler
[locker.compiler] Compiling 303 files to /Users/jason/code/scala2/build/locker/classes/compiler
[locker.compiler] Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=384m; support was removed in 8.0
[locker.compiler] /Users/jason/code/scala2/src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:444: warning: could not emit switch for @switch annotated match
[locker.compiler]       (const.tag: @switch) match {
[locker.compiler]                    ^
[locker.compiler] warning: there were 169 deprecation warnings; re-run with -deprecation for details
[locker.compiler] warning: there were 17 unchecked warnings; re-run with -unchecked for details
[locker.compiler] warning: there were 262 inliner warnings; re-run with -Yinline-warnings for details
[locker.compiler] four warnings found
[propertyfile] Creating new property file: /Users/jason/code/scala2/build/locker/classes/compiler/compiler.properties
     [copy] Copying 5 files to /Users/jason/code/scala2/build/locker/classes/compiler
[stopwatch] [locker.compiler.timer: 1:22.784 sec]

locker.done:

quick.start:

quick.lib:
    [mkdir] Created dir: /Users/jason/code/scala2/build/quick/classes/library
    [javac] Compiling 32 source files to /Users/jason/code/scala2/build/quick/classes/library
    [javac] warning: [options] bootstrap class path not set in conjunction with -source 1.5
    [javac] warning: [options] source value 1.5 is obsolete and will be removed in a future release
    [javac] warning: [options] To suppress warnings about obsolete options, use -Xlint:-options.
    [javac] 3 warnings
[quick.library] Compiling 579 files to /Users/jason/code/scala2/build/quick/classes/library
[quick.library] Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=384m; support was removed in 8.0
[quick.library] error: scala.reflect.internal.FatalError: object ScalaRunTime does not have a member hash
[quick.library]     at scala.reflect.internal.Definitions$DefinitionsClass.scala$reflect$internal$Definitions$DefinitionsClass$$fatalMissingSymbol(Definitions.scala:1181)
[quick.library]     at scala.reflect.internal.Definitions$DefinitionsClass.getMember(Definitions.scala:1198)
[quick.library]     at scala.tools.nsc.backend.jvm.ScalacBackendInterface.<init>(ScalacBackendInterface.scala:111)
[quick.library]     at scala.tools.nsc.backend.icode.GenICode.<init>(GenICode.scala:34)
[quick.library]     at scala.tools.nsc.Global$genicode$.<init>(Global.scala:585)
[quick.library]     at scala.tools.nsc.Global.genicode$lzycompute(Global.scala:581)
[quick.library]     at scala.tools.nsc.Global.genicode(Global.scala:581)
[quick.library]     at scala.tools.nsc.Global.computeInternalPhases(Global.scala:703)
[quick.library]     at scala.tools.nsc.Global.computePhaseDescriptors(Global.scala:735)
[quick.library]     at scala.tools.nsc.Global.phaseDescriptors$lzycompute(Global.scala:742)
```

retronym committed:
Avoid initialization order problem

retronym commented:
Here's a fix:

https://github.com/retronym/scala/commit/ac75d108392e45704822c29fbd1fa1906c806a19

You must not call into `Definitions` before `Definitions#init()` has been called.

DarkDimius committed:
Fixing bugs introduced while porting.

ArrayClass is magical. Several methods had recursive implementation. PatternMatching was shadowing parameter.

DarkDimius committed:
Adapt unit tests to use ScalacBackendInterface

DarkDimius committed:
Scala requires compilation unit to generate fresh symbol

DarkDimius commented:
@retronym thanks for help!

I tried to run all testsuite with modified genbcode. I have failures for tests such as pos/t5545, but it seems that I'm bug compatible with https://issues.scala-lang.org/browse/SI-8431
@lrytz Are you up to review? Should I squash commits or would it be easier for you review them separately?

DarkDimius committed:
It seems that for tests order of definitions inside Definitions matters.

DarkDimius commented:
The left failures are one neg and one run. I can reproduce neither of those. I can compile failed tests by hand with self-bootstrapped compiler, while it seems that in Jenkins they fail with NPE. I don't have a stacktrace to see where this NPE was issued, so I'm not sure how to fix it.

retronym commented:
The test failure was just a friendly reminder you either to regenerate the source of `JavaUniverseForce` or to restore the previous order of the members of `Definition`. It didn't mean that semantics of the compiler were affected by that ordering.

retronym commented:
@DarkDimius I am also unable to locally reproduce the [test failures](https://scala-webapps.epfl.ch/jenkins/job/pr-scala-test/7509/console) in

```
neg/case-collision2.scala
test/files/run/t8925.scala
```

After building a merge of d37ad and `origin/2.11.x`.

Let's see if it is deterministic:

PLS REBUILD/pr-scala@d37adac374f4127f66ff74e8cac756f6e0f74665

scala-jenkins commented:
(kitty-note-to-self: ignore 63910525)
:cat: Roger! Rebuilding pr-scala for d37adac3. :rotating_light:

sjrd commented:
OMG this is definitely going to endanger the cross-compilation of the Scala.js backend against all versions of Scala ...
I will try to compile Scala.js against this branch, and see how much breaks, and whether I can come up with [more implicit magic](https://github.com/scala-js/scala-js/blob/master/compiler/src/main/scala/scala/scalajs/compiler/Compat210Component.scala) to survive this.

Note that I'm not even speaking about actually using that abstraction. At first glance, at least the changes in `ScalaPrimitives` are not source compatible.

DarkDimius commented:
@sjrd 
There are indeed changes to ScalaPrimitives and Opcodes that are not backwards source compatible. But those changes allow sharing those classes between compilers.
While they are not backwards source compatible it is pretty easy to come up way to adapt code so that it compiles in both old previous versions  and new ones. One could supply dummy objects for scalaPrimitives and opcodes so that both 
`import opcodes._` and `import Opcodes._` will typecheck, but actual definition will code from either depending on actual scalac version that you compile against.

DarkDimius commented:
@retronym those failures seem to be deterministic: they were there in all commits in this branch.

DarkDimius committed:
Move in shared implementation between dotty and scalac.

Those methods ended having same implementation in dotty and scalac.
And even more they are implemented in terms of what is already part of exposed abstractions.
Providing default implementations, that can be overridden in backends if ever needed.

DarkDimius committed:
Workaround name-based pattern matching creating tuples for ValDefs.

See comment here: https://issues.scala-lang.org/browse/SI-8989?focusedCommentId=71196&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-71196
 for illustration.

DarkDimius committed:
Fix error in BCodeBodyBuilder.

Found by just looking at the code. Interestingly enough none of the tests broke.

sjrd commented:
I get these two compilation warning when compiling your branch (tip: 2d0104dbb344bce442f65a0af14d6e1d1158e2e8):

```
quick.comp:
[quick.compiler] Compiling 36 files to /localhome/doeraene/projects/scala/build/quick/classes/compiler
[quick.compiler] /localhome/doeraene/projects/scala/src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:446: warning: could not emit switch for @switch annotated match
[quick.compiler]       (const.tag: @switch) match {
[quick.compiler]                    ^
[quick.compiler] /localhome/doeraene/projects/scala/src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:622: warning: matches with two cases or fewer are emitted using if-then-else instead of switch
[quick.compiler]               (argsSize : @switch) match {
[quick.compiler]                            ^
```

These warnings do not appear when compiling 2.11.x.

sjrd commented:
OK it appears I have managed to come up with the appropriate magic to maintain source compatibility in our codebase:
https://github.com/scala-js/scala-js/pull/1307

DarkDimius commented:
It just became apparent to me that name-based pattern matching based approach doesn't work with shared instances in scalac: 
Unlike in dotty where code such as:

```
tree match {
      case Block(stats, expr) =>
         doSmth
         expr
}
```

will be rewritten to something similar to

```
val obj = Block.unapply(tree)
val stats = obj._1
val expr = obj._2
doSmth
expr
```

in scalac it's rewritten to 

```
val obj = Block.unapply(tree)
doSmth[stats -> obj._1, expr -> obj._2]
obj._2
```

where [a -> b] means substitution.

The scheme used by scala means that one cannot reuse same objects for name-based pattern matching, as actual extractor call are postponed, and if `doSmth` in this example matches on `Block` in the same manner it will rewrite value stored in obj._2.

I guess this is a bug in patmat in terms that it should capture values for name-based pattern matching as those are defined by defs and can change

DarkDimius commented:
Created an issue for this: https://issues.scala-lang.org/browse/SI-9003

DarkDimius commented:
So, to make it work and work fast we:
1. cannot use name-based pattern matching due to https://issues.scala-lang.org/browse/SI-9003
2. cannot use name-based pattern matching for Product1
3. cannot use type test, without inheriting from same interface both in scalac and dotty, as ClassTags  create options for every test. Solution could be-replace type tests by call to methods that return booleans, or trees themselves implement same interfaces for dotty and scala.
4. Methods that aren't part of minimal interface, in order to have good performance require implicit helpers to provide implementation. Usage of same share same returned objects seems to not create any problems now, but it could if somebody will start writing code not taking this into account.

I believe that in the end the potential solution requires to many workarounds and is too ugly to merge.

DarkDimius commented:
It seems that this PR revealed regression from 2.11.4 that current master has.
See https://issues.scala-lang.org/browse/SI-9003 for details.
Is it possible to configure jenkins to use released 2.11.4 to build this branch?

retronym committed:
SI-9003 Eagerly capture more potentially mutable binders

This is a re-run of SI-5158 in two different contexts.

As reported, the result of `unapply[Seq]` under name based pattern
matching might not guarantee stability of results of calls to
`_1`, `apply(i)`, etc, so we call these eagerly, and call them
once only.

I also found a case in regular pattern matching that we hadn't
accounted for: extracting elements of sequences (either from
a case class or from an `unapplySeq`) may also be unstable.

This commit changes `ExtractorTreeMaker` to force storage
of such binders, even under `-optimize`. This parallels the change
to `ProductExtractorTreeMaker` in 8ebe8e3e8.

I have added a special case for traditional `unapply` methods
returning `Option`. This avoids a change for:

```
% cat test/files/run/t9003b.scala
object Single {
  def unapply(a: Any) = Some("")
}

object Test {
  def main(args: Array[String]): Unit = {
    "" match {
      case Single(x) =>
        (x, x)
    }
  }
}
% qscalac -optimize -Xprint:patmat test/files/run/t9003b.scala 2>&1 | grep --context=5 get
      case <synthetic> val x1: Any = "";
      case5(){
        <synthetic> val o7: Some[String] = Single.unapply(x1);
        if (o7.isEmpty.unary_!)
          matchEnd4({
            scala.Tuple2.apply[String, String](o7.get, o7.get);
            ()
          })
        else
          case6()
      };
% scalac-hash v2.11.4 -optimize -Xprint:patmat test/files/run/t9003b.scala 2>&1 | grep --context=5 get
      case <synthetic> val x1: Any = "";
      case5(){
        <synthetic> val o7: Some[String] = Single.unapply(x1);
        if (o7.isEmpty.unary_!)
          matchEnd4({
            scala.Tuple2.apply[String, String](o7.get, o7.get);
            ()
          })
        else
          case6()
      };
```

DarkDimius commented:
depends on #4164

DarkDimius committed:
Get rid of most of allocations in ScalacBackendInterface

By reusing same decorator instances.

DarkDimius committed:
Fix several bugs introduced by capturing a sub pattern instead of performing a test.

DarkDimius committed:
Make reading of source easier by giving a bit more structure.

Also gather in one place Product1s

DarkDimius commented on src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:
```diff
@@ -1218,40 +1235,40 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
         val equalsMethod: Symbol = {
           if (l.tpe <:< BoxedNumberClass.tpe) {
-            if (r.tpe <:< BoxedNumberClass.tpe) platform.externalEqualsNumNum
-            else if (r.tpe <:< BoxedCharacterClass.tpe) platform.externalEqualsNumChar
```
this should call `BoxesRuntime.equalsNumChar`. But this method is private(unlike all other equals-like methods). How could it call it?

lrytz replied:
very good question ;-) https://github.com/scala/scala/blob/2.11.x/src/library/scala/runtime/BoxesRunTime.java#L186

```
lucmac:scala2 luc$ scala
Welcome to Scala version 2.11.4 (Java HotSpot(TM) 64-Bit Server VM, Java 1.6.0_65).
Type in expressions to have them evaluated.
Type :help for more information.

scala> def foo(a: java.lang.Number, b: java.lang.Character) = a == b
scala.reflect.internal.FatalError: object BoxesRunTime does not have a decl equalsNumChar
    at scala.reflect.internal.Definitions$DefinitionsClass.scala$reflect$internal$Definitions$DefinitionsClass$$fatalMissingSymbol(Definitions.scala:1179)
    at scala.reflect.internal.Definitions$DefinitionsClass.getDecl(Definitions.scala:1253)
    at scala.tools.nsc.backend.JavaPlatform$class.externalEqualsNumChar(JavaPlatform.scala:41)
    at scala.tools.nsc.Global$GlobalPlatform.externalEqualsNumChar$lzycompute(Global.scala:96)
    at scala.tools.nsc.Global$GlobalPlatform.externalEqualsNumChar(Global.scala:96)
    at scala.tools.nsc.backend.icode.GenICode$ICodePhase.genEqEqPrimitive(GenICode.scala:1498)
```

https://github.com/scala/scala/pull/4189

DarkDimius committed:
There's no support for early initialisers in dotty. Make initialisation more lazy.

DarkDimius committed:
Handle Dotty Idents

In Scalac any function call has Select as it function. In dotty it could be an Ident.
Eg in Scalac you get Select(Preded$, println);
while in dotty you get Ident(println)
In order to get actual receiver need to either look at the type(which is TermRef) or at owner chain.
First one is more systematic but is isn't exposed in the abstraction API, so using the second one.

DarkDimius committed:
Handle dotty typeApply.

TypeApply's are free to not have a Apply node around them in dotty.

DarkDimius committed:
Refactor genArrayValue, so that array creation from elements could be reused

DarkDimius committed:
Minimal set of modifications to allow to support both dotty <label> DefDefs and scalac LabelDefs

DefDefs are assumed to be grouped in blocks that share same return address used to jump back.

DarkDimius committed:
A bit more abstraction on what are primitives and operations on them.

Required to support arrays in dotty. Operations on those couldn't be decided based on symbol.

DarkDimius committed:
Remove allocations of Options also for Product1 extractors.

DarkDimius commented:
note for reviewers: CompactLocalVariablesTest  fails due to #4164

DarkDimius committed:
Handle Dotty Closures. Generate InvokeDynamic for them. Give a basis to generate them for Scalac

Based on https://github.com/retronym/scala/compare/topic/invoke-dynamic-lambda adapted to abstract from global.

DarkDimius committed:
Give types and positions to all symbols that backend creates.

DarkDimius committed:
Address reviewers comments: less casts needed.

DarkDimius committed:
Make difference between java private flag and scala private flag.

DarkDimius commented:
GenBcode doesn't acutally use anything in Primitives that has to do with TypeKinds. I've splitted Primitives to a trait and an object, only object part is used by GenBcode. No need for type projections anymore.

adriaanm commented:
I'm reviewing the blocking PR #4164 now.

adriaanm commented:
Merged.

DarkDimius committed:
Use 2.11.5-SNAPSHOT for star until 2.11.5 is released.

DarkDimius committed:
Dotty specific: handle lading of This in types.

DarkDimius committed:
Uniform handling of scalac ArrayValue and JavaSeqLiteral in dotty.

DarkDimius committed:
Handle Selects inside closures.

DarkDimius committed:
Initial generation of default methods.

lrytz committed:
snapshot scala

lrytz commented:
@DarkDimius: pull in this diff to enable `2.11.5-SNAPSHOT` as starr: https://github.com/lrytz/scala/commit/8628f927d1366389f9ec7821ff53ae27a2bc5424

lrytz commented on src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:
```diff
@@ -795,19 +833,56 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
+
+      if(!int.shouldEmitJumpAfterLabels) genNormalBlock
+      else {
+        val (prefixLabels: List[LabelDef], stats1) = stats.span {
```
```
[locker.compiler] /Users/luc/scala/scala2/src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:857: warning: abstract type LabelDef in type pattern List[BCodeBodyBuilder.this.int.LabelDef] (the underlying of List[BCodeBodyBuilder.this.int.LabelDef]) is unchecked since it is eliminated by erasure
[locker.compiler]         val (prefixLabels: List[LabelDef], stats1) = stats.span {
[locker.compiler]                            ^
```

DarkDimius committed:
Merge pull request #1 from lrytz/sharedBackend

snapshot scala

DarkDimius committed:
Merge branch 'sharing-backend' of github.com:DarkDimius/scala into sharing-backend

lrytz commented on src/compiler/scala/tools/nsc/backend/jvm/BackendInterface.scala:
```diff
@@ -0,0 +1,793 @@
+  val LongTag: ConstantTag
+  val StringTag: ConstantTag
+  val ClazzTag: ConstantTag
+  val EnumTag: ConstantTag
```
This abstraction breaks the [`@switch` in `genConstant`](https://github.com/scala/scala/blob/2.11.x/src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala#L439):

```
[quick.compiler] /Users/luc/scala/scala2/src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:459: warning: could not emit switch for @switch annotated match
[quick.compiler]       (const.tag: @switch) match {
[quick.compiler]                    ^
```

Not specific to this change: We need to do at least some performance testing to get a feeling for the performance penalty before adding the new abstraction layer.

DarkDimius replied:
For this particular abstraction, actual ConstantTag==Long values are same in dotty and scalac, so if this particular place introduces a slowdown it could be solved.
The way how I observe how fast is the compiler is time that is taken to compile standard library. Before 2d22839 it was 10% slower. After I didn't check.

DarkDimius committed:
Handle dotty Idents by desugaring them into selects.

Handling Ident's in dotty requires analysing its type,
Instead of exporting a lot of abstractions around types
I'd better go with a single method that hides this.

lrytz commented on src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:
```diff
@@ -522,55 +546,82 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
+         */
+        for (i <- args.length until dims) elemKind = ArrayBType(elemKind)
+      }
+      (argsSize : @switch) match {
```
can you remove the `@swtich` here? it triggers a warning:

```
[quick.compiler] /Users/luc/scala/scala2/src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:606: warning: matches with two cases or fewer are emitted using if-then-else instead of switch
[quick.compiler]       (argsSize : @switch) match {
[quick.compiler]                    ^
```

lrytz commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -632,7 +631,9 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
     val runsRightAfter = None
-  } with GenBCode
+  } with GenBCode {
+    lazy val int = new ScalacBackendInterface[global.type](global)
```
when looking through the entire diff of this PR, there's 4 instances of `new ScalacBackendInterface` - one ought to be enough, no?

lrytz commented on src/compiler/scala/tools/nsc/backend/icode/ICodeCheckers.scala:
```diff
@@ -520,23 +522,23 @@ abstract class ICodeCheckers {
            checkStack(instr.consumed)
            primitive match {
-             case Negation(kind) =>
+             case Negation(kind: TypeKind) =>
```
do we need those type annotations?

lrytz commented on src/compiler/scala/tools/nsc/backend/icode/ICodes.scala:
```diff
@@ -22,14 +22,15 @@ abstract class ICodes extends AnyRef
                                  with TypeKinds
-                                 with ExceptionHandlers
                                  with Primitives
+                                 with ExceptionHandlers
```
is there a reason for this change?

lrytz commented on src/compiler/scala/tools/nsc/backend/icode/Opcodes.scala:
```diff
@@ -307,30 +308,30 @@ trait Opcodes { self: ICodes =>
-        case Conversion(from, _)   => from :: Nil
-        case ArrayLength(kind)     => ARRAY(kind) :: Nil
-        case StringConcat(kind)    => ConcatClass :: kind :: Nil
+        case Negation(kind: TypeKind)        => kind :: Nil
```
type annotations necessary?

lrytz commented on src/compiler/scala/tools/nsc/backend/icode/analysis/TypeFlowAnalysis.scala:
```diff
@@ -204,16 +208,16 @@ abstract class TypeFlowAnalysis {
         case CALL_PRIMITIVE(primitive) =>
           primitive match {
-            case Negation(kind) => stack.pop; stack.push(kind)
+            case Negation(kind: TypeKind) => stack.pop; stack.push(kind)
```
type annotations

lrytz commented on src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala:
```diff
@@ -32,7 +26,7 @@ final class BCodeAsmCommon[G <: Global](val global: G) {
     // always top-level. However, SI-8900 shows an example where the weak name-based implementation
     // of isDelambdafyFunction failed (for a function declared in a package named "lambda").
-    classSym.isAnonymousClass || !classSym.originalOwner.isClass
+    classSym.isAnonymousClass || (classSym.originalOwner != NoSymbol && !classSym.originalOwner.isClass)
```
why this change?

lrytz commented on src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:
```diff
@@ -351,13 +368,20 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
-            else { locals.load(sym) }
-            generatedType = tk
+          val tk = symInfoTK(sym)
+          generatedType = tk
```
i'm not conviced it's safe to move this assignment out of the `if`. it might change the behavior, because at the end, `genLoad` checks `if (generatedType != expectedType) ...`.

lrytz commented on src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:
```diff
@@ -378,14 +402,18 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
         case EmptyTree => if (expectedType != UNIT) { emitZeroOf(expectedType) }
 
+
+        case t: TypeApply => // dotty specific
```
dotty-specific: see https://github.com/DarkDimius/scala/commit/9392c811dc9f563d0ef2ece46cbeaed03cb5ff3b#commitcomment-8851371

DarkDimius committed:
Even more handling of primitives by trees.

needed to handle array_clone in dotty. Which doesn't have a symbol but isn't byte code primitive operation

lrytz commented on src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala:
```diff
@@ -637,7 +644,7 @@ abstract class GenASM extends SubComponent with BytecodeWriters { self =>
     def descriptor(s: Symbol):   String = { javaType(s).getDescriptor }
 
-    def javaType(tk: TypeKind): asm.Type = {
+    def javaType(tk: TypeKinds#TypeKind): asm.Type = {
```
are these projections in GenASM still necessary?

lrytz commented on src/compiler/scala/tools/nsc/backend/jvm/BackendInterface.scala:
```diff
@@ -0,0 +1,795 @@
+    def exists(pred: Tree => Boolean): Boolean
+  }
+  
+  trait SymbolHelper {
```
As I mentioned before, I'm not so happy that symbols get new extension methods in the backend. I think we should restrict the methods here to those that are actually members of class Symbol. For the others (like `getsJavaPrivateFlag` etc) I'd prefer if they were just ordinary methods that take the symbol as parameter.

For the cases when a method is a member on scalac's symbol class, but not in dotty's, i think it would be better to opt for an ordinary method too, instead of an extension method in the `DottyBackendInterface`, for the same reason.

DarkDimius committed:
Merge branch 'sharing-backend' of github.com:DarkDimius/scala into sharing-backend

lrytz commented:
Let's close this one for now, as it requires a chunk of work.
